include/dlaf/permutations/general/impl.h

```diff
@@ -0,0 +1,201 @@
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include "dlaf/permutations/general/api.h"
+
+#include "dlaf/blas/tile.h"
+#include "dlaf/common/index2d.h"
+#include "dlaf/lapack/gpu/lacpy.h"
+#include "dlaf/lapack/tile.h"
+#include "dlaf/matrix/copy_tile.h"
+#include "dlaf/matrix/matrix.h"
+#include "dlaf/schedulers.h"
+#include "dlaf/sender/transform.h"
+#include "dlaf/sender/when_all_lift.h"
+#include "dlaf/types.h"
+#include "dlaf/util_matrix.h"
+
+#include "pika/parallel/algorithms/for_each.hpp"
+
+namespace dlaf::permutations::internal {
+
+// Applies the permutaton index `perm_arr` to a portion of the columns/rows(depends on coord) [1] of an
+// input submatrix [2] and saves the result into a subregion [3] of an output submatrix [4].
+//
+// Example column permutations with `perm_arr = [8, 2, 5]`:
+//
+//          2     5     8      out_begin
+//     ┌────────────────────┐    ┌─────┬─────────────┐
+//     │  in_offset      in │    │     │         out │
+//     │ │                  │    │     └─►┌─┬─┬─┐    │
+//     │◄┘ ┌─┐   ┌─┐   ┌─┐  │    │        │c│a│b│    │
+//     │   │a│   │b│   │c│  │    │    ┌──►│ │ │ │    │
+//     │   │ │ ┌►│ │   │ │◄─┼────┼──┐ │   └─┴─┴─┘    │
+//     │   └─┘ │ └─┘   └─┘  │    │  │ │      ▲       │
+//     │       │            │    │sz.rows()  │       │
+//     │      sz.rows()     │    │           └─      │
+//     │                    │    │          sz.cols()│
+//     └────────────────────┘    └───────────────────┘
+//
+// Example row permutations with `perm_arr = [3, 1]`:
+//
+//             ┌─── in_offset
+//             │                   out_begin
+//     ┌───────▼────────────┐    ┌──┬────────────────┐
+//     │                 in │    │  │            out │
+//     │       ┌─────┐      │    │  └►┌─────┐        │
+//   1 │       │  a  │      │    │    │  b  │        │
+//     │       └─────┘      │    │    ├─────┤ ◄─┐    │
+//     │                    │    │    │  a  │   │    │
+//     │       ┌─────┐      │    │    └──▲──┘   │    │
+//   3 │       │  b  │      │    │       │      │    │
+//     │       └──▲──┘      │    │ sz.cols()    │    │
+//     │          │         │    │          sz.rows()│
+//     └──────────┼─────────┘    └───────────────────┘
+//                │
+//             sz.cols()
+//
+//
+// [1]: The portion of each input column or row is defined by the interval [in_offset, in_offset +
+//      sz.col()) or the interval [in_offset, in_offset + sz.row()) respectively.
+// [2]: The input submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `in_tiles`
+// [3]: The subregion is defined by `begin` and `sz`
+// [4]: The output submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `out_tiles`
+//
+// Note: `in_tiles` should be `const T` but to avoid extra allocations necessary for unwrapping
```

Comment:
Note, I don't think this would lead to additional allocations, but definitely more clunky usage.

Reply:
To unwrap `std::vector<pika::shared_future<matrix::Tile<const T, D>>>` we need to do something like:

```
template <class U>
std::vector<matrix::Tile<const U, Device::CPU>> unwrapConstTile(
    const std::vector<pika::shared_future<matrix::Tile<const U, Device::CPU>>>& tiles_fut) {
  std::vector<matrix::Tile<const U, Device::CPU>> tiles;
  tiles.reserve(tiles_fut.size());
  for (const auto& tile_fut : tiles_fut) {
    const auto& tile = tile_fut.get();
    SizeType len = tile.size().isEmpty() ? 0 : tile.ld() * (tile.size().cols() - 1) + tile.size().rows();
    tiles.emplace_back(tile.size(), MemoryView(tile.ptr(), len), tile.ld());
  }
  return tiles;
}
```

so I guess you are write in a sense because there is no new heap memory allocated from `MemoryView`. The allocations for the `std::vector<matrix::Tile<const T, D>>` is not a problem.

Reply:
Ah, well, yes the `vector` allocation is indeed a new allocation. I was imagining it being possible to work directly with the `shared_future::get` references from the first vector avoiding the second vector completely. As I said, more clunky so I wouldn't do it, but I imagine it should be possible.

TL;DR: I'm only commenting on the comment. I have no problems with the decision to use non-const `T`.